contracts/USM.sol

```diff
@@ -1,117 +1,131 @@
 pragma solidity ^0.6.7;
 
-import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import "@openzeppelin/contracts/math/SafeMath.sol";
-import "./oracles/IOracle.sol";
+import "@openzeppelin/contracts/math/Math.sol";
+import "./BufferedToken.sol";
+import "./WadMath.sol";
+import "./FUM.sol";
 import "@nomiclabs/buidler/console.sol";
 
+
 /**
  * @title USM Stable Coin
  * @author Alex Roan (@alexroan)
- * @notice Concept by Jacob Eliosoff (@jacob-eliosoff)
+ * @notice Concept by Jacob Eliosoff (@jacob-eliosoff).
  */
-contract USM is ERC20 {
+contract USM is BufferedToken {
     using SafeMath for uint;
+    using Math for uint;
+    using WadMath for uint;
+
+    uint constant public MIN_ETH_AMOUNT = WAD / 1000;         // 0.001 ETH
+    uint constant public MIN_BURN_AMOUNT = WAD;               // 1 USM
+    uint constant public MAX_DEBT_RATIO = 900000000000000000; // 90%
 
-    uint constant UNIT = 1e27; // 10**27, MakerDAO's RAY
-    uint constant MINT_FEE = UNIT / 1000; // 0.1%
-    uint constant BURN_FEE = UNIT / 200; // 0.5%
+    FUM public fum;
+    uint public latestFumPrice;
 
-    address oracle;
-    uint public ethPool;
+    event LatestFumPriceChanged(uint previous, uint latest);
 
     /**
      * @param _oracle Address of the oracle
      */
-    constructor(address _oracle) public ERC20("Minimal USD", "USM") {
-        oracle = _oracle;
-        ethPool = 0;
-        console.log("Hi!");
-        console.log(_oracle);
+    constructor(address _oracle) public BufferedToken(_oracle, "Minimal USD", "USM") {
+        _setLatestFumPrice(MIN_ETH_AMOUNT);
+        fum = new FUM();
     }
 
     /**
-     * @notice Multiplies x and y, assuming they are both fixed point with 27 digits.
+     * @notice Calculates the price of FUM using its total supply
+     * and ETH buffer
      */
-    function mulFixed(uint256 x, uint256 y) internal pure returns (uint256) {
-        return x.mul(y).div(UNIT);
-    }
+    function fumPrice() public view returns (uint) {
+        int buffer = ethBuffer();
+        // if we're underwater, use the last fum price
+        if (buffer <= 0 || debtRatio() > MAX_DEBT_RATIO) {
+            return latestFumPrice;
+        }
 
-    /**
-     * @notice Divides x by y, assuming they are both fixed point with 27 digits.
-     */
-    function divFixed(uint256 x, uint256 y) internal pure returns (uint256) {
-        return x.mul(UNIT).div(y);
+        uint fumTotalSupply = fum.totalSupply().max(WAD); // fumTotalSupply is floored at WAD
+        return uint(buffer).wadDiv(fumTotalSupply);
```

Comment:
I suggest something like:

```
theoreticalFumPrice = uint(buffer).wadDiv(fumTotalSupply);
minFumPrice = _ethToUsm(_minFumBuyPriceInEth);
if (minFumPrice > theoreticalFumPrice) {    // Does Solidity have a max() fn?  Maybe we should write one...
    return minFumPrice;
} else {
    return theoreticalFumPrice;
}

Reply:
We are essentially doing that a few lines above but with `latestFumPrice` instead of `_minFumBuyPrice`.

Reply:
I see your point, but in fact (again, per my possibly-improvable algo) we need to compare theoretical price to min price, not just check whether we're underwater.  Because there are cases where although we're still underwater, the theoretical price climbs **higher** than the min price, so buyers should pay the former.  Example:

1. 100 ETH @ $400 ($40,000), 30,000 USM (buffer $10,000, dr 75%), 5,000 FUM @ $2
2. Price $400 -> $200.  Debt ratio crossed 80%, mfbp calculated as: (1-80%) * 100 / 5,000 = 0.004 ETH:
100 ETH @ $200 ($20,000), 30,000 USM (buffer -$10,000, dr 150%), 5,000 FUM @ -$2 (min 0.004 ETH = $0.80)
3. Fund 50:
150 ETH @ $200 ($30,000), 30,000 USM (buffer $0, dr 100%), 5,000 FUM @ $0 (min 0.004 ETH = $0.80)
4. Price $200 -> $240:
150 ETH @ $240 ($36,000), 30,000 USM (buffer $6,000, dr 83.33%), 5,000 FUM @ **$1.20** (min 0.004 ETH = **$0.96**)